gc/gc-base/src/main/java/org/projectnessie/gc/base/IdentifiedResultsRepo.java

```diff
@@ -0,0 +1,202 @@
+   * @return spark dataset of row where each row is having the expired contents per content id per
+   *     reference.
+   */
+  public Dataset<Row> collectExpiredContentsAsDataSet(String runId) {
+    // Example Query:
+    // SELECT * FROM nessie.db1.`identified_results@someGcRef`
+    //    WHERE gcRunId = '9e809dc9-ac35-41c9-b1c0-5ebcf452ea2d'
+    return sql(
+        "SELECT * FROM %s WHERE %s = '%s'",
+        catalogAndTableWithRefName,
+        //
+        COL_GC_RUN_ID,
+        runId);
+  }
+
+  public String getLatestCompletedRunID() {
+    // collect row for the last written run id
+    // Example query:
+    // SELECT gcRunId FROM nessie.db2.`identified_results@someGcBranch` WHERE gcRunStart =
+    //    (SELECT MAX(gcRunStart) FROM nessie.db2.`identified_results@someGcBranch`) LIMIT 1
+    List<Row> rows =
+        sql(
+                "SELECT %s FROM %s WHERE %s = (SELECT MAX(%s) FROM %s) LIMIT 1",
+                COL_GC_RUN_ID,
+                //
+                catalogAndTableWithRefName,
+                //
+                COL_GC_RUN_START,
+                COL_GC_RUN_START,
+                catalogAndTableWithRefName)
+            .collectAsList();
+    return rows.isEmpty() ? null : rows.get(0).getString(0);
+  }
+
+  void writeToOutputTable(Dataset<Row> rowDataset) {
+    try {
+      // write content rows to the output table
+      rowDataset.writeTo(catalogAndTableWithRefName).append();
+    } catch (NoSuchTableException e) {
+      throw new RuntimeException(e);
+    }
+  }
+
+  static Row createContentRow(
+      Content content, String runId, Timestamp startedAt, long snapshotId, Reference ref) {
+    return RowFactory.create(
+        startedAt,
+        runId,
+        content.getId(),
+        content.getType().name(),
+        snapshotId,
+        ref.getName(),
+        ref.getHash());
+  }
+
+  private void createTableIfAbsent(
+      SparkSession sparkSession,
+      String catalogName,
+      TableIdentifier tableIdentifier,
+      String gcRefName) {
+    // get Nessie catalog
+    Catalog nessieCatalog =
+        CatalogUtil.loadCatalog(
+            NessieCatalog.class.getName(),
+            catalogName,
+            catalogConfWithRef(sparkSession, catalogName, gcRefName),
+            sparkSession.sparkContext().hadoopConfiguration());
+    try {
+      nessieCatalog.createTable(tableIdentifier, icebergSchema);
+    } catch (AlreadyExistsException ex) {
+      // Table can exist from previous GC run, no need to throw exception.
+    }
+  }
+
+  private static String withRefName(String catalog, String identifier, String refName) {
+    int tableNameIndex = identifier.lastIndexOf(".");
+    String namespace = identifier.substring(0, tableNameIndex);
+    String tableName = identifier.substring(tableNameIndex + 1);
+    return catalog
+        + "."
+        + namespace
+        + "."
+        + ImmutableTableReference.builder().name(tableName).reference(refName).build();
```

Comment:
so we are relying on the auto-generated `toString` method for the `ImmutableTableReference`. Is it safe?

Reply:
`TableReference` has a `toString() `override. Immutables will use that.